File: tests/keytest.py
#!/usr/bin/env python3
# Сверка имён плиток: то, что печатает приложение (tests/keytest), против
# того, что даёт нарезалка. Подробности — в шапке tests/keytest.cpp.
import importlib.util, math, sys, os

HERE = os.path.dirname (os.path.abspath (__file__))
spec = importlib.util.spec_from_file_location (
           "tiler", os.path.join (HERE, "..", "tools", "dwd-tiles.py"))
tiler = importlib.util.module_from_spec (spec)
spec.loader.exec_module (tiler)

bad = n = 0
for line in sys.stdin:
    parts = line.split()
    if len (parts) != 4:
        continue
    tile, lon, lat, key = int (parts[0]), float (parts[1]), float (parts[2]), parts[3]
    mine = tiler.tile_key (math.floor (lon/tile)*tile, math.floor (lat/tile)*tile)
    n += 1
    if mine != key:
        bad += 1
        if bad <= 8:
            print (f"  [FAIL] плитка {tile}°, {lon},{lat} — "
                   f"приложение {key}, нарезалка {mine}")

print (f"  [{' OK ' if bad == 0 else 'FAIL'}] имена плиток сходятся "
       f"-> сверено {n}, расхождений {bad}")
sys.exit (1 if bad or n == 0 else 0)

File: tests/keytest.cpp
// Имена плиток по обе стороны договора.
//
// Нарезалка (tools/dwd-tiles.py) даёт файлу имя по юго-западному углу
// квадрата, а приложение считает то же имя само, по углам экрана, и по
// нему просит файл. Разойдись они хоть на ведущем нуле — приложение
// искало бы то, чего нет, и молча показывало «в этом районе пусто».
// Ошибка без единого сообщения, самая дорогая порода.
//
// Программа печатает имена, а сверяет их с питоновской стороной
// tests/keytest.py: держать два независимых счётчика в одном языке
// смысла нет, договор именно межъязыковой.
#include <cstdio>
#include <clocale>
#include <QCoreApplication>
#include "DwdTiles.h"

int main (int argc, char **argv)
{
	QCoreApplication app (argc, argv);
	// Qt при создании QCoreApplication переводит локаль на системную, и
	// printf начинает печатать дробные с запятой. Разбор на той стороне
	// на этом ломается — уже наступали.
	setlocale (LC_ALL, "C");

	for (int tile : {5, 20})
		for (double lat = -85; lat <= 85; lat += 2.5)
			for (double lon = -180; lon < 180; lon += 2.5)
				printf ("%d %.4f %.4f %s\n", tile, lon, lat,
				        DwdTiles::key (lon, lat, tile).toUtf8().constData());
	return 0;
}
